File: pico8.py
"""
PICO-8 cart support for Tiled
2020, <sam@hocevar.net>
"""
import tiled as T
import base64

PALETTE = [
    (0x00, 0x00, 0x00), # black
    (0x1d, 0x2b, 0x53), # dark_blue
    (0x7e, 0x25, 0x53), # dark_purple
    (0x00, 0x87, 0x51), # dark_green
    (0xab, 0x52, 0x36), # brown
    (0x5f, 0x57, 0x4f), # dark_gray
    (0xc2, 0xc3, 0xc7), # light_gray
    (0xff, 0xf1, 0xe8), # white
    (0xff, 0x00, 0x4d), # red
    (0xff, 0xa3, 0x00), # orange
    (0xff, 0xec, 0x27), # yellow
    (0x00, 0xe4, 0x36), # green
    (0x29, 0xad, 0xff), # blue
    (0x83, 0x76, 0x9c), # indigo
    (0xff, 0x77, 0xa8), # pink
    (0xff, 0xcc, 0xaa), # peach

    (0x29, 0x18, 0x14),
    (0x11, 0x1d, 0x35),
    (0x42, 0x21, 0x36),
    (0x12, 0x53, 0x59),
    (0x74, 0x2f, 0x29),
    (0x49, 0x33, 0x3b),
    (0xa2, 0x88, 0x79),
    (0xf3, 0xef, 0x7d),
    (0xbe, 0x12, 0x50),
    (0xff, 0x6c, 0x24),
    (0xa8, 0xe7, 0x2e),
    (0x00, 0xb5, 0x43),
    (0x06, 0x5a, 0xb5),
    (0x75, 0x46, 0x65),
    (0xff, 0x6e, 0x59),
    (0xff, 0x9d, 0x81),
]

class PICO8(T.Plugin):
    @classmethod
    def nameFilter(self):
        return "PICO-8 cart (*.p8)"

    @classmethod
    def shortName(self):
        return "pico8"

    @classmethod
    def supportsFile(self, f):
        return open(f, 'rb').read(16) == b'pico-8 cartridge'

    @classmethod
    def read(self, filename):
        with open(filename, 'rb') as f:
            cart = f.read()

        def extract(header):
            pos = cart.find(header)
            if pos < 0:
                return ''
            pos += len(header)
            data = cart[pos:cart.find(b'_',pos)]
            return ''.join(data.decode('ascii').split())

        # Create a map
        m = T.Tiled.Map(T.Tiled.Map.Orthogonal, 128, 64, 8, 8)
        m.setBackgroundColor(T.qt.QColor(*PALETTE[0]))
        m.setProperty('data', base64.b64encode(cart))

        # Create an image and a tileset for the palette
        pal = [T.qt.QColor(*rgb).rgb() for rgb in PALETTE]
        tsize = 12
        img = T.qt.QImage(4 * tsize, 8 * tsize, T.qt.QImage.Format_Indexed8)
        img.setColorTable(pal)
        img.fill(0)
        for n in range(32):
            x, y = n % 4 * tsize, n // 4 * tsize
            for j in range(tsize):
                for i in range(tsize):
                    img.setPixel(x + i, y + j, n)
        t = T.Tiled.Tileset.create('PICO-8 Palette', tsize, tsize, 0, 0)
        t.data().loadFromImage(img, '')
        m.addTileset(t)

        # Read gfx data into an image
        gfxdata = extract(b'__gfx__')
        pal = [T.qt.QColor(*rgb).rgb() for rgb in PALETTE]
        img = T.qt.QImage(128, 128, T.qt.QImage.Format_Indexed8)
        img.setColorTable(pal)
        img.fill(0)
        for i in range(min(128*128, len(gfxdata))):
            if i % 128 < 8 and i // 128 < 8:
                continue
            img.setPixel(i % 128, i // 128, int(gfxdata[i], 16))

        # Create fileset from the image
        t = T.Tiled.Tileset.create('PICO-8 Sprites', 8,8, 0, 0)
        t.data().loadFromImage(img, '')
        m.addTileset(t)

        # Read map data into a layer
        mapdata = extract(b'__map__')
        l = T.Tiled.TileLayer('PICO-8 Map', 0, 0, 128, 64)
        for i in range(min(128*64, len(mapdata) // 2)):
            n = int(mapdata[i*2:i*2+2], 16)
            if n > 0:
                ti = t.data().tileAt(n)
                l.setCell(i % 128, i // 128, T.Tiled.Cell(ti))

        # 2nd half of gfx data can overlap 2nd half of map
        for i in range(128*64, min(128*128, len(gfxdata)), 2):
            n = int(gfxdata[i:i+2][::-1], 16)
            if n > 0:
                ti = t.data().tileAt(n)
                l.setCell(i // 2 % 128, i // 256, T.Tiled.Cell(ti))

        m.addLayer(l)

        return m

    @classmethod
    def write(self, m, filename):

        try:
            cart = base64.b64decode(m.propertyAsString('data'))
        except:
            raise Exception('This map was not loaded from a PICO-8 cart.')

        eol = b'\r\n' if cart.find(b'\r\n') >= 0 else b'\n'

        def split(header):
            gpos = cart.find(header)
            if gpos < 0:
                gpos = len(cart)
            gend = cart.find(b'_', gpos + 7)
            if gend < 0:
                gend = len(cart)
            data = ''.join(cart[gpos+7:gend].decode('ascii').split())
            return cart[:gpos], data, cart[gend:]

        def chunk(s, n):
            empty = b'0' * n
            while s[-n:] == empty:
                s = s[:-n] # Remove trailing empty lines
            return [s[i:i+n] for i in range(0, len(s), n)]

        l = T.tileLayerAt(m, 0)

        # Insert gfx data into cart
        prefix, data, suffix = split(b'__gfx__')

        # First 64 lines of gfx data are from the original cart; add padding
        # if not enough data in there
        newdata = data[:128*64]
        newdata += ''.join(['0' for x in range(128*64-len(newdata))])

        # The next 64 lines actually come from the last 32 lines of the map
        for n in range(128*32, 128*64):
            t = l.cellAt(n % 128, n // 128).tile()
            tid = t.id() if t else 0
            newdata += '{0:02x}'.format(tid)[::-1]

        # Convert chunked data to bytes
        newdata = eol.join([b'__gfx__'] + chunk(newdata.encode('ascii'), 128) + [b''])
        cart = prefix + newdata + suffix

        # Insert map data into cart
        prefix, _, suffix = split(b'__map__')

        newdata = ''
        for n in range(128*32):
            t = l.cellAt(n % 128, n // 128).tile()
            tid = t.id() if t else 0
            newdata += '{0:02x}'.format(tid)

        # Conert chunked data to bytes
        newdata = eol.join([b'__map__'] + chunk(newdata.encode('ascii'), 256) + [b''])
        cart = prefix + newdata + suffix

        with open(filename, 'wb') as f:
            f.write(cart)

        return True

File: pico8.js
//
//  PICO-8 cart support for Tiled
//
//  Copyright © 2020—2023 Sam Hocevar <sam@hocevar.net>
//
//  This program is free software. It comes without any warranty, to
//  the extent permitted by applicable law. You can redistribute it
//  and/or modify it under the terms of the Do What the Fuck You Want
//  to Public License, Version 2, as published by the WTFPL Task Force.
//  See http://www.wtfpl.net/ for more details.
//

const PALETTE =
[
    '#000000', // black
    '#1d2b53', // dark_blue
    '#7e2553', // dark_purple
    '#008751', // dark_green
    '#ab5236', // brown
    '#5f574f', // dark_gray
    '#c2c3c7', // light_gray
    '#fff1e8', // white
    '#ff004d', // red
    '#ffa300', // orange
    '#ffec27', // yellow
    '#00e436', // green
    '#29adff', // blue
    '#83769c', // indigo
    '#ff77a8', // pink
    '#ffccaa', // peach

    '#291814',
    '#111d35',
    '#422136',
    '#125359',
    '#742f29',
    '#49333b',
    '#a28879',
    '#f3ef7d',
    '#be1250',
    '#ff6c24',
    '#a8e72e',
    '#00b543',
    '#065ab5',
    '#754665',
    '#ff6e59',
    '#ff9d81',
];

const TILED_VERSION = tiled.version.split('.').map((e,i) => e*100**(2-i)).reduce((a,b) => a+b);
const HEADER = 'pico-8 cartridge';
const PROPNAME = 'Private Data';

const MAP_WIDTH = 128;
const MAP_HEIGHT = 64;

function tohex(x, ndigits)
{
    return (x + (1 << (ndigits * 4))).toString(16).slice(-ndigits);
}

function fromhex(s)
{
    return Number('0x'+s);
}

// Extract a hexadecimal section from a p8 cart data, e.g. ‘__gfx__’
function p8_extract(buf, header)
{
    return p8_split(buf, header)[1];
}

// Split a p8 cart in order to replace a hex section
function p8_split(buf, header)
{
    let gpos = buf.indexOf(header);
    gpos = gpos < 0 ? buf.length : gpos + header.length;
    let gend = buf.indexOf('__', gpos);
    if (gend < 0)
        gend = buf.length;
    return [ buf.slice(0, gpos),
             buf.slice(gpos, gend).replace(/[^0-9a-fA-F]+/g, ''),
             buf.slice(gend, buf.length) ];
}

function pico8_read(filename)
{
    let f = new BinaryFile(filename);
    let cart = f.readAll().toString();
    f.close();

    if (cart.slice(0, HEADER.length) != HEADER)
        throw new TypeError('Not a PICO-8 cartridge!');

    // Create a map
    let tm = new TileMap();
    tm.setSize(MAP_WIDTH, MAP_HEIGHT);
    tm.setTileSize(8, 8);
    tm.orientation = TileMap.Orthogonal;
    tm.backgroundColor = PALETTE[0];
    //tm.setProperty('Show Sprite 0', false);
    tm.setProperty(PROPNAME, Qt.btoa(cart));

    // Create an image and a tileset for the palette
    let tsize = 12;
    // TODO

    // Read gfx data into an image
    let gfx = p8_extract(cart, '__gfx__');
    let img = new Image(128, 128, Image.Format_Indexed8);
    img.setColorTable(PALETTE);
    for (let i = 0; i < Math.min(128 * 128, gfx.length); ++i)
        img.setPixel(i % 128, Math.floor(i / 128), fromhex(gfx[i]));

    // Create a tileset from sprite image
    let t = new Tileset('PICO-8 Sprites');
    t.backgroundColor = PALETTE[3];
    t.setTileSize(8, 8);
    t.loadFromImage(img);
    tm.addTileset(t);

    // Read map data into a tile layer
    let map = p8_extract(cart, '__map__');
    let tl = new TileLayer();
    tl.width = MAP_WIDTH;
    tl.height = MAP_HEIGHT;
    let tle = tl.edit();
    function set_tile(x, y, s)
    {
        let id = fromhex(s);
        if (id > 0)
            tle.setTile(x, y, t.tile(id));
    }
    for (let i = 0; i < Math.min(MAP_HEIGHT * MAP_WIDTH, Math.floor(map.length / 2)); ++i)
        set_tile(i % MAP_WIDTH,
                 Math.floor(i / MAP_WIDTH),
                 map.substring(i * 2, i * 2 + 2));
    // The second part of the sprite data also contains map data
    let gfx2 = gfx.slice(128 * 64).replace(/(.)(.)/g, '$2$1');;
    for (let i = 0; i < Math.min(128 * 64, gfx2.length); i += 2)
        set_tile(Math.floor(i / 2) % MAP_WIDTH,
                 MAP_HEIGHT / 2 + Math.floor(i / 2 / MAP_WIDTH),
                 gfx2.substring(i, i + 2));
    tle.apply();
    tm.addLayer(tl);

    return tm;
}

function pico8_write(tm, filename)
{
    let cart = Qt.atob(tm.property(PROPNAME));
    if (cart.slice(0, HEADER.length) != HEADER)
        throw new TypeError('This map was not loaded from a PICO-8 cart');

    let layer = tm.layerAt(0);
    let eol = cart.indexOf('\r\n') >= 0 ? '\r\n' : '\n';

    // Convert map data to hex
    let data = '';
    for (let i = 0; i < MAP_WIDTH * MAP_HEIGHT; ++i)
    {
        let t = layer.cellAt(i % MAP_WIDTH, Math.floor(i / MAP_WIDTH)).tileId;
        data += tohex(Math.max(t, 0), 2);
    }

    // Retrieve gfx data from the original cart.
    let [ prefix, gfx, suffix ] = p8_split(cart, '__gfx__');

    // Load the tileset used
    let tileset = tm.tilesets[0];
    let gfxOut = '';

    // If there's an image set, the user has loaded an external tileset which we
    // now need to store in the cart (on next load, it'll be pulled back out of
    // the cart)
    if (tileset.image)
    {
        // The first 64 lines are generated from the tileset image
        let tilesetImage = new Image(tileset.image);

        // Only take the first 128x64 pixels worth
        for (let i = 0; i < 128 * 64; ++i)
        {
            // But allow the image itself to be any shape
            let p = tilesetImage.pixelColor(i % tilesetImage.width, Math.floor(i / tilesetImage.width));
            // Assume the image is already formatted in the PICO-8 palette
            let c = PALETTE.findIndex((color) => color === p.toString().toLowerCase())
            // Write it out as color 0 if there's not an exact match found
            // TODO: Find the nearest colour instead
            gfxOut += tohex(Math.max(c, 0), 1);
        }
    }
    else
    {
        // The first 64 lines are preserved from the .p8 file
        gfxOut = gfx.slice(0, 128 * 64).padEnd(128 * 64, '0');
    }

    // The next 64 lines are taken from the map section because they are from the
    // shared area.
    gfxOut = gfxOut + data.slice(128 * 64, 128 * 128).replace(/(.)(.)/g, '$2$1');

    // Remove empty lines and store
    gfxOut = gfxOut.slice(0, 128 * 128).replace(/(0{128})+$/, '');
    cart = [prefix].concat(gfxOut.match(/.{128}/g)).concat(suffix).join(eol);

    // Store map data. Contrary to gfx data, nothing is preserved.
    [ prefix, map, suffix ] = p8_split(cart, '__map__');
    map = data.slice(0, 256 * 32).replace(/(0{256})+$/, '');
    cart = [prefix].concat(map.match(/.{256}/g)).concat(suffix).join(eol);

    // Save the file
    let f = new BinaryFile(filename, BinaryFile.WriteOnly);
    f.write(cart);
    f.commit();
}

if (TILED_VERSION >= 10500)
{
    const pico8_format =
    {
        name: 'PICO-8 cart (*.p8)',
        extension: 'p8',
        read: pico8_read,
        write: pico8_write,
    };

    tiled.registerMapFormat('PICO-8', pico8_format);
}
else
{
    console.warn(`Tiled version ${tiled.version} is too old for the PICO-8 plugin (1.5.0 required)`);
}
